common/pulp_docker/common/models.py

```diff
@@ -170,6 +173,48 @@ def __init__(self, digest, name, tag, architecture, fs_layers, history, schema_v
                 "The DockerManifest class only supports Docker v2, Schema 1 manifests.")
         self.schema_version = schema_version
 
+    @staticmethod
+    def digest(manifest):
+        """
+        Calculate the sha256 digest of the Manifest, given in JSON.
+
+        :param manifest: The raw JSON representation of the Manifest.
+        :type  manifest: basestring
+        :return:         The digest of the given Manifest
+        :rtype:          basestring
+        """
+        decoded_manifest = json.loads(manifest, encoding='utf-8')
+        if 'signatures' in decoded_manifest:
+            # This manifest contains signatures. Unfortunately, the Docker manifest digest
+            # is calculated on the unsigned version of the Manifest so we need to remove the
+            # signatures. To do this, we will look at the 'protected' key within the first
+            # signature. This key indexes a (malformed) base64 encoded JSON dictionary that
+            # tells us how many bytes of the manifest we need to keep before the signature
+            # appears in the original JSON. We will strip out the bytes after this cutoff point,
+            # add back a trailing newline and closing curly brace, and then calculate the sha256
+            # sum of the transformed JSON to get the digest.
+            protected = decoded_manifest['signatures'][0]['protected']
+            # Docker has not included the required padding at the end of the base64 encoded
+            # 'protected' block. We need to add the correct number of ='s signs back.
+            protected = protected.replace('\n', '').replace(' ', '')
```

Comment:
I never saw manifests where the protected value contained these characters, but the Docker code also does this same thing so I figured it was there for a reason:

https://github.com/docker/libtrust/blob/9cbd2a1374f46905c68a4eb3694a130610adc62a/util.go#L167-L168